packages/types/src/type/ExtrinsicEra.ts

```diff
@@ -2,33 +2,180 @@
 // This software may be modified and distributed under the terms
 // of the Apache-2.0 license. See the LICENSE file for details.
 
+import EnumType from '../codec/EnumType';
+import { isHex, isU8a, hexToU8a, isObject, assert, u8aToBn } from '@polkadot/util';
+import Tuple from '../codec/Tuple';
+import U8 from '../primitive/U8';
 import { AnyU8a } from '../types';
-
-import { u8aToU8a } from '@polkadot/util';
-
 import U8a from '../codec/U8a';
 
 /**
  * @name ExtrinsicEra
  * @description
  * The era for an extrinsic, indicating either a mortal or immortal extrinsic
  */
-export default class ExtrinsicEra extends U8a {
+export default class ExtrinsicEra extends EnumType<ImmortalEra | MortalEra> {
+  constructor (value?: any, index?: number) {
+    super({ ImmortalEra, MortalEra }, value, index);
+  }
+
+  /**
+   * @description Returns the item as a [[ImmortalEra]]
+   */
+  get asImmortalEra (): ImmortalEra {
+    assert(this.isImmortalEra, `Cannot convert '${this.type}' via asImmortalEra`);
+    return this.value as ImmortalEra;
+  }
+
+  /**
+   * @description Returns the item as a [[MortalEra]]
+   */
+  get asMortalEra (): MortalEra {
+    assert(this.isMortalEra, `Cannot convert '${this.type}' via asMortalEra`);
+    return this.value as MortalEra;
+  }
+
+  /**
+   * @description `true` if Immortal
+   */
+  get isImmortalEra (): boolean {
+    return this.index === 0;
+  }
+
+  /**
+   * @description `true` if Mortal
+   */
+  get isMortalEra (): boolean {
+    return this.index === 1;
+  }
+
+  /**
+   * @description Encodes the value as a Uint8Array as per the parity-codec specifications
+   * @param isBare true when the value has none of the type-specific prefixes (internal)
+   */
+  toU8a (isBare?: boolean): Uint8Array {
+    if (this.index === 0) {
+      return super.toU8a();
+    } else {
+      return this.asMortalEra.toU8a(isBare);
+    }
+  }
+}
+
+const VALID_IMMORTAL = new U8a([0]);
+/**
+ * @name ImmortalEra
+ * @description
+ * The ImmortalEra for an extrinsic
+ */
+export class ImmortalEra extends U8a {
   constructor (value?: AnyU8a) {
-    super(
-      ExtrinsicEra.decodeExtrinsicEra(value)
-    );
+    super(value);
+
+    assert(this.eq(VALID_IMMORTAL), `IMMORTAL: expected ${VALID_IMMORTAL.toHex()}, found ${this.toHex()}`);
```

Comment:
can you try `this.eq([0])` if this works, then you can just let VALID_IMMORTAL = [0]